File: neuro-san-client/src/neuro_san_client/agent_runner.py

# Copyright © 2023-2026 Cognizant Technology Solutions Corp, www.cognizant.com.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
#
# END COPYRIGHT
"""
agent_runner: the BYOK chat-loop driver used by both the Python and TS clients.

For the MVP, this delegates orchestration to the **origin's** /streaming_chat
endpoint. The client supplies LLM credentials via `sly_data.llm_config`; the
origin uses them per-request without persisting. This means:

    - the client doesn't need a JS/Python LLM SDK,
    - cross-origin agents resolve at the origin via existing ExternalActivation,
    - client-side coded tools (`client_side_source`) are NOT exercised in this
      MVP path (the origin runs everything). Adding them is a follow-up.

Mirror: neuro-san-lite-js/src/agent_runner.ts
"""
from __future__ import annotations

import json
from dataclasses import dataclass, field
from typing import Any, AsyncIterator, Dict, Optional
from urllib.parse import urlparse

import httpx

from . import wire_config


@dataclass
class ChatEvent:
    """One event from the chat stream, surfaced to the UI."""

    kind: str    # "user" | "agent" | "thinking" | "network" | "error" | "done"
    payload: Any


@dataclass
class RunnerOptions:
    """Per-turn options."""

    url: str
    message: str
    anthropic_key: str = ""
    openai_key: str = ""
    chat_context: Dict[str, Any] = field(default_factory=dict)
    sly_data: Dict[str, Any] = field(default_factory=dict)
    timeout_seconds: float = 180.0


@dataclass
class RunnerResult:
    """Final state at the end of a turn — useful for the caller to thread back in."""

    chat_context: Dict[str, Any] = field(default_factory=dict)
    sly_data: Dict[str, Any] = field(default_factory=dict)
    answer: str = ""


# ---------------------------------------------------------------------------
# Internals: small helpers shared by the streaming runner.
# ---------------------------------------------------------------------------


def _derive_streaming_chat_url(network_url: str, network_name: str) -> str:
    """
    Given a notebook URL like https://host/api/v1/{name}/network, derive the
    matching /streaming_chat endpoint. We use the wire-config-stamped origin
    so a misleadingly-shaped network_url cannot redirect us off-origin.
    """
    parsed = urlparse(network_url)
    if not parsed.scheme or not parsed.netloc:
        raise ValueError(f"network_url is not fully qualified: {network_url!r}")
    return f"{parsed.scheme}://{parsed.netloc}/api/v1/{network_name}/streaming_chat"


def _build_request(opts: RunnerOptions) -> Dict[str, Any]:
    """Build the streaming_chat request body, threading the BYOK keys into
    sly_data.llm_config."""
    sly_data: Dict[str, Any] = dict(opts.sly_data or {})

    # Thread BYOK keys into sly_data.llm_config so the server's
    # replace_any_required_api_keys() picks them up.
    llm_keys: Dict[str, Any] = dict(sly_data.get("llm_config") or {})
    if opts.anthropic_key:
        llm_keys.setdefault("api_key", opts.anthropic_key)
        llm_keys["anthropic_api_key"] = opts.anthropic_key
    if opts.openai_key:
        llm_keys.setdefault("api_key", opts.openai_key)
        llm_keys["openai_api_key"] = opts.openai_key
    if llm_keys:
        sly_data["llm_config"] = llm_keys

    request: Dict[str, Any] = {
        "user_message": {"type": "HUMAN", "text": opts.message},
        # MAXIMAL: pass intermediate AGENT messages too (default MINIMAL
        # drops them). We need them for the network_call events that
        # ExternalActivation embeds in the stream — without them, the
        # browser can't render cross-origin choreography in its trace panel.
        "chat_filter": {"chat_filter_type": "MAXIMAL"},
    }
    if opts.chat_context:
        request["chat_context"] = opts.chat_context
    if sly_data:
        request["sly_data"] = sly_data
    return request


# ---------------------------------------------------------------------------
# Public API.
# ---------------------------------------------------------------------------


async def fetch_notebook(url: str,
                          client: Optional[httpx.AsyncClient] = None) -> Dict[str, Any]:
    """Fetch and JSON-decode the Agent Web notebook at `url`.

    Tests inject an httpx.AsyncClient backed by httpx.MockTransport.
    """
    owns_client = client is None
    if owns_client:
        client = httpx.AsyncClient(timeout=30.0)
    try:
        resp = await client.get(url)
        resp.raise_for_status()
        return resp.json()
    finally:
        if owns_client:
            await client.aclose()


async def run_agent_turn(opts: RunnerOptions,
                          client: Optional[httpx.AsyncClient] = None
                          ) -> AsyncIterator[ChatEvent]:
    """
    Drive one chat turn.

    Yields ChatEvent objects as the conversation streams. The final event is
    always a ChatEvent("done", RunnerResult).

    Tests inject a client that uses httpx.MockTransport for deterministic
    network behavior.
    """
    async for event in _run_agent_turn_impl(opts, client):
        yield event


async def _run_agent_turn_impl(opts: RunnerOptions,
                                client: Optional[httpx.AsyncClient]
                                ) -> AsyncIterator[ChatEvent]:
    owns_client = client is None
    if owns_client:
        client = httpx.AsyncClient(timeout=opts.timeout_seconds)
    result = RunnerResult(
        chat_context=dict(opts.chat_context or {}),
        sly_data=dict(opts.sly_data or {}),
    )

    try:
        # 1) Fetch notebook.
        yield ChatEvent("network", {"kind": "notebook", "method": "GET",
                                     "url": opts.url, "status": None})
        try:
            resp = await client.get(opts.url)
        except httpx.HTTPError as exc:
            yield ChatEvent("error", f"Notebook fetch failed: {exc}")
            yield ChatEvent("done", result)
            return
        if resp.status_code != 200:
            yield ChatEvent("error",
                            f"Notebook GET {opts.url} returned HTTP {resp.status_code}")
            yield ChatEvent("done", result)
            return
        yield ChatEvent("network", {"kind": "notebook", "method": "GET",
                                     "url": opts.url, "status": resp.status_code})
        notebook = resp.json()

        # 2) Extract + verify wire config.
        try:
            wire = wire_config.extract_wire_config_from_notebook(notebook)
            wire_config.verify_wire_config(wire)
        except wire_config.WireConfigError as exc:
            yield ChatEvent("error", f"Wire config invalid: {exc}")
            yield ChatEvent("done", result)
            return

        # 3) Stream the chat against the origin.
        network_name = wire_config.get_network_name(wire)
        streaming_url = _derive_streaming_chat_url(opts.url, network_name)
        request_body = _build_request(opts)

        yield ChatEvent("network", {"kind": "streaming_chat", "method": "POST",
                                     "url": streaming_url, "status": None})

        compiled_parts: list = []

        try:
            async with client.stream(
                "POST", streaming_url,
                content=json.dumps(request_body).encode("utf-8"),
                headers={"Content-Type": "application/json"},
            ) as stream_resp:
                if stream_resp.status_code != 200:
                    body = await stream_resp.aread()
                    yield ChatEvent(
                        "error",
                        f"streaming_chat HTTP {stream_resp.status_code}: "
                        f"{body[:300]!r}"
                    )
                    yield ChatEvent("done", result)
                    return
                yield ChatEvent("network", {"kind": "streaming_chat",
                                             "method": "POST", "url": streaming_url,
                                             "status": stream_resp.status_code})

                async for line in stream_resp.aiter_lines():
                    if not line:
                        continue
                    try:
                        chunk = json.loads(line)
                    except json.JSONDecodeError:
                        continue
                    response = chunk.get("response") or {}
                    msg_type = response.get("type")
                    text = response.get("text")
                    chat_ctx = response.get("chat_context")
                    sly = response.get("sly_data")
                    structure = response.get("structure") or {}
                    if chat_ctx:
                        result.chat_context = chat_ctx
                    if sly:
                        result.sly_data.update(sly)
                    # Server-side cross-origin calls (made by ExternalActivation
                    # or RemoteToolActivation upstream of us) embed a
                    # `network_call` field in the message's structure. Surface
                    # those as network events so a runtime's trace panel can
                    # render the choreography the origin executed on the
                    # caller's behalf.
                    net_call = structure.get("network_call") if isinstance(structure, dict) else None
                    if net_call:
                        # Mark which origin reported the call so the UI can
                        # distinguish browser-direct calls from server-reported
                        # ones.
                        annotated = dict(net_call)
                        annotated.setdefault("via", opts.url)
                        yield ChatEvent("network", annotated)
                    # AGENT_FRAMEWORK carries the front man's final user-facing
                    # answer; AI is direct LLM output (usually filtered out by
                    # default chat_filter); AGENT / AGENT_PROGRESS /
                    # AGENT_TOOL_RESULT are intermediate "thinking" steps.
                    if msg_type in ("AGENT_FRAMEWORK", "AI") and text:
                        compiled_parts.append(text)
                        yield ChatEvent("agent", text)
                    elif msg_type in ("AGENT", "AGENT_PROGRESS",
                                       "AGENT_TOOL_RESULT") and text:
                        yield ChatEvent("thinking", text)
        except httpx.HTTPError as exc:
            yield ChatEvent("error", f"streaming_chat failed: {exc}")
            yield ChatEvent("done", result)
            return

        result.answer = "".join(compiled_parts)
        yield ChatEvent("done", result)
    finally:
        if owns_client:
            await client.aclose()

File: neuro-san-lite-js/src/agent_runner.ts
// Port of neuro_san_client/agent_runner.py
// BYOK chat-loop driver. Streams events out as an async iterable so the UI
// can render incrementally.

import type {
    ChatEvent,
    Json,
    RunnerOptions,
    RunnerResult,
} from "./types.js";
import {
    extractWireConfigFromNotebook,
    getNetworkName,
    verifyWireConfig,
    WireConfigError,
} from "./wire_config.js";

/** Derive the /streaming_chat URL for a given /network URL and network name. */
export function deriveStreamingChatUrl(networkUrl: string, networkName: string): string {
    let parsed: URL;
    try {
        parsed = new URL(networkUrl);
    } catch {
        throw new Error(`networkUrl is not fully qualified: ${networkUrl}`);
    }
    if (!parsed.protocol || !parsed.host) {
        throw new Error(`networkUrl is not fully qualified: ${networkUrl}`);
    }
    return `${parsed.protocol}//${parsed.host}/api/v1/${networkName}/streaming_chat`;
}

/**
 * Build the streaming_chat request body, threading BYOK keys into
 * sly_data.llm_config. Caller-supplied keys take precedence over BYOK ones,
 * matching the Python behavior (setdefault).
 */
export function buildRequest(opts: RunnerOptions): Record<string, Json> {
    // Start with caller-supplied sly_data (don't mutate it).
    const slyData: Record<string, Json> = { ...(opts.slyData ?? {}) };

    // Merge BYOK keys into sly_data.llm_config without overriding caller's choices.
    const existingLlm = (slyData["llm_config"] as Record<string, Json> | undefined) ?? {};
    const llmKeys: Record<string, Json> = { ...existingLlm };

    if (opts.anthropicKey) {
        if (llmKeys["api_key"] === undefined) {
            llmKeys["api_key"] = opts.anthropicKey;
        }
        llmKeys["anthropic_api_key"] = opts.anthropicKey;
    }
    if (opts.openaiKey) {
        if (llmKeys["api_key"] === undefined) {
            llmKeys["api_key"] = opts.openaiKey;
        }
        llmKeys["openai_api_key"] = opts.openaiKey;
    }
    if (Object.keys(llmKeys).length > 0) {
        slyData["llm_config"] = llmKeys;
    }

    const request: Record<string, Json> = {
        user_message: { type: "HUMAN", text: opts.message },
        // MAXIMAL: pass intermediate AGENT messages too (default MINIMAL
        // drops them). We need them for the network_call events that
        // ExternalActivation embeds in the stream — without them, the
        // browser can't render cross-origin choreography in its trace panel.
        chat_filter: { chat_filter_type: "MAXIMAL" },
    };
    if (opts.chatContext && Object.keys(opts.chatContext).length > 0) {
        request["chat_context"] = opts.chatContext as Json;
    }
    if (Object.keys(slyData).length > 0) {
        request["sly_data"] = slyData;
    }
    return request;
}

/**
 * Drive one chat turn. Yields ChatEvent objects as the conversation streams.
 * The final event is always `{kind: "done", payload: RunnerResult}`.
 */
export async function* runAgentTurn(opts: RunnerOptions): AsyncGenerator<ChatEvent, void, void> {
    const fetchImpl = opts.fetchImpl ?? fetch;
    const result: RunnerResult = {
        chatContext: { ...(opts.chatContext ?? {}) },
        slyData: { ...(opts.slyData ?? {}) },
        answer: "",
    };

    // 1) Fetch notebook.
    yield {
        kind: "network",
        payload: { kind: "notebook", method: "GET", url: opts.url, status: null },
    };
    let response: Response;
    try {
        response = await fetchImpl(opts.url);
    } catch (exc) {
        yield { kind: "error", payload: `Notebook fetch failed: ${(exc as Error).message}` };
        yield { kind: "done", payload: result };
        return;
    }
    if (!response.ok) {
        yield {
            kind: "error",
            payload: `Notebook GET ${opts.url} returned HTTP ${response.status}`,
        };
        yield { kind: "done", payload: result };
        return;
    }
    yield {
        kind: "network",
        payload: {
            kind: "notebook",
            method: "GET",
            url: opts.url,
            status: response.status,
        },
    };
    let notebook: unknown;
    try {
        notebook = await response.json();
    } catch (exc) {
        yield { kind: "error", payload: `Notebook JSON parse failed: ${(exc as Error).message}` };
        yield { kind: "done", payload: result };
        return;
    }

    // 2) Extract + verify wire config.
    let wire;
    try {
        wire = extractWireConfigFromNotebook(notebook);
        verifyWireConfig(wire);
    } catch (exc) {
        const msg = exc instanceof WireConfigError
            ? `Wire config invalid: ${exc.message}`
            : `Wire config invalid: ${(exc as Error).message}`;
        yield { kind: "error", payload: msg };
        yield { kind: "done", payload: result };
        return;
    }

    // 3) Stream the chat against the origin.
    const networkName = getNetworkName(wire);
    const streamingUrl = deriveStreamingChatUrl(opts.url, networkName);
    const requestBody = buildRequest(opts);

    yield {
        kind: "network",
        payload: {
            kind: "streaming_chat",
            method: "POST",
            url: streamingUrl,
            status: null,
        },
    };

    let streamResp: Response;
    try {
        streamResp = await fetchImpl(streamingUrl, {
            method: "POST",
            headers: { "Content-Type": "application/json" },
            body: JSON.stringify(requestBody),
        });
    } catch (exc) {
        yield { kind: "error", payload: `streaming_chat failed: ${(exc as Error).message}` };
        yield { kind: "done", payload: result };
        return;
    }
    if (!streamResp.ok) {
        const body = await streamResp.text();
        yield {
            kind: "error",
            payload: `streaming_chat HTTP ${streamResp.status}: ${body.slice(0, 300)}`,
        };
        yield { kind: "done", payload: result };
        return;
    }
    yield {
        kind: "network",
        payload: {
            kind: "streaming_chat",
            method: "POST",
            url: streamingUrl,
            status: streamResp.status,
        },
    };

    const compiledParts: string[] = [];
    for await (const line of streamLines(streamResp)) {
        if (!line) {
            continue;
        }
        let chunk: { response?: ChunkResponse };
        try {
            chunk = JSON.parse(line);
        } catch {
            continue;
        }
        const respObj = chunk?.response ?? {};
        const msgType = respObj.type;
        const text = respObj.text;
        const chatCtx = respObj.chat_context;
        const sly = respObj.sly_data;
        const structure = respObj.structure;
        if (chatCtx) {
            result.chatContext = chatCtx;
        }
        if (sly) {
            Object.assign(result.slyData, sly);
        }
        // Server-side cross-origin calls (ExternalActivation / RemoteToolActivation
        // upstream of us) embed a `network_call` field in the message's
        // structure. Surface those as network events so the trace panel can
        // render the choreography the origin executed on the caller's behalf.
        if (structure && typeof structure === "object" && structure.network_call) {
            const annotated = {
                ...structure.network_call,
                via: structure.network_call["via"] ?? opts.url,
            };
            yield { kind: "network", payload: annotated };
        }
        // AGENT_FRAMEWORK carries the front man's final user-facing answer;
        // AI is direct LLM output (usually filtered out by default chat_filter);
        // AGENT / AGENT_PROGRESS / AGENT_TOOL_RESULT are intermediate "thinking"
        // steps that we surface separately.
        const finalKinds = new Set(["AGENT_FRAMEWORK", "AI"]);
        const thinkingKinds = new Set(["AGENT", "AGENT_PROGRESS", "AGENT_TOOL_RESULT"]);
        if (typeof msgType === "string" && finalKinds.has(msgType) && typeof text === "string" && text) {
            compiledParts.push(text);
            yield { kind: "agent", payload: text };
        } else if (typeof msgType === "string" && thinkingKinds.has(msgType) && typeof text === "string" && text) {
            yield { kind: "thinking", payload: text };
        }
    }

    result.answer = compiledParts.join("");
    yield { kind: "done", payload: result };
}

/** Shape of one chunk on the streaming_chat wire. */
interface ChunkResponse {
    type?: string;
    text?: string;
    chat_context?: Record<string, Json>;
    sly_data?: Record<string, Json>;
    structure?: {
        network_call?: {
            kind?: string;
            method?: string;
            url?: string;
            status?: number | null;
            ms?: number;
            via?: string;
        };
        [k: string]: unknown;
    };
}

/**
 * Yield successive lines from a streaming Response body. Handles chunked
 * transfer where line boundaries don't align with read chunks. Tolerates both
 * \n and \r\n line endings.
 */
export async function* streamLines(response: Response): AsyncGenerator<string, void, void> {
    if (!response.body) {
        // Some test fakes have no body; treat as empty.
        return;
    }
    const reader = response.body.getReader();
    const decoder = new TextDecoder("utf-8");
    let buffer = "";
    try {
        while (true) {
            const { value, done } = await reader.read();
            if (done) {
                break;
            }
            buffer += decoder.decode(value, { stream: true });
            let newlineIdx: number;
            while ((newlineIdx = buffer.indexOf("\n")) >= 0) {
                let line = buffer.slice(0, newlineIdx);
                buffer = buffer.slice(newlineIdx + 1);
                if (line.endsWith("\r")) {
                    line = line.slice(0, -1);
                }
                yield line;
            }
        }
        // Flush any trailing line without a terminator.
        buffer += decoder.decode();
        if (buffer.length > 0) {
            yield buffer;
        }
    } finally {
        reader.releaseLock();
    }
}
